repository-s3/src/main/java/io/aiven/elasticsearch/repositories/s3/S3OutputStream.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2020 Aiven Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.elasticsearch.repositories.s3;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.model.AbortMultipartUploadRequest;
+import com.amazonaws.services.s3.model.CompleteMultipartUploadRequest;
+import com.amazonaws.services.s3.model.InitiateMultipartUploadRequest;
+import com.amazonaws.services.s3.model.ObjectMetadata;
+import com.amazonaws.services.s3.model.PartETag;
+import com.amazonaws.services.s3.model.UploadPartRequest;
+import org.elasticsearch.common.Strings;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class S3OutputStream extends OutputStream {
+
+    private final Logger logger = LoggerFactory.getLogger(S3OutputStream.class);
+
+    private final AmazonS3 client;
+
+    private final ByteBuffer byteBuffer;
+
+    private final String bucketName;
+
+    private final String key;
+
+    private MultipartUpload multipartUpload;
+
+    private final int partSize;
+
+    public S3OutputStream(final String bucketName,
+                           final String key,
+                           final int partSize,
+                           final AmazonS3 client) {
+        this.bucketName = bucketName;
+        this.key = key;
+        this.client = client;
+        this.partSize = partSize;
+        this.byteBuffer = ByteBuffer.allocate(partSize);
+    }
+
+    @Override
+    public void write(final int b) throws IOException {
+        write(new byte[]{(byte) b}, 0, 1);
+    }
+
+    @Override
+    public void write(final byte[] b, final int off, final int len) throws IOException {
+        if (Objects.isNull(b) || b.length == 0) {
+            return;
+        }
+        if (Objects.isNull(multipartUpload)) {
+            multipartUpload = newMultipartUpload();
+        }
+        final var source = ByteBuffer.wrap(b, off, len);
```

Comment:
Instead of manual buffering, please consider using `BufferedOutputStream` (e.g. internally).

Reply:
As we discuss due to the `BufferedOutputStream::write` methond better to use so far the custom solution